.gitlab-ci.yml

```diff
@@ -32,6 +32,17 @@ build:
     # Clear *_TOKEN variables during code build so that compile-time code can't access them
     - apt-get update && apt-get install -y --no-install-recommends libpq-dev git rsync
     - CI_BUILD_TOKEN="" KUBE_TOKEN="" PROD_KUBE_TOKEN="" PROD_DOCKER_PASSWORD="" etc/scripts/stage_docker.sh --install-ghc --test
+    - |
+      if [[ "$CI_BUILD_REF_NAME" == "master" ]]; then
+        aws s3 cp etc/docker/_artifacts/curator s3://${CI_S3_BUCKET}/curator/curator-latest-linux-amd64
```

Comment:
This looks really good and meets the requirements set out in the issue but I'd like to suggest that this is also probably a good time to update the CI to better organize the jobs and stages here. 

So what that entails is splitting up the `build` job we currently have into more isolated and compartmentalized jobs (still within the build "stage" though).

In Gitlab stages make it handy to define how different separate jobs can be run in parallel within the same stage.

From the docs https://docs.gitlab.com/ee/ci/yaml/#stage

> stage is defined per-job and relies on stages which is defined globally. It allows to group jobs into different stages, and jobs of the same stage are executed in parallel (subject to certain conditions). 

So here is a perfect situation and time to start breaking up the CI into different stages.

With the changes being added/made for this issue, we should consider this now since we're now doing several things in the `build` stage of the CI (and they now target slightly different end goals).

1. We want to have several separate jobs in the `build` stages. 
    So before where we just had one job (`build`) in the `build` stage and it was just used in the CI to build and push the docker image that the later `deploy_prod` and `deploy_ci` jobs used in the deploy stage. 

    But now would be a great time to separate out the task that produces a dedicated binary that we will upload into a separate job in the `build`  stage. 

2. The s3 copy portion of the CI should go into a separate job in the `deploy` stage. So we'll have a job that depends on the binary-build job in the `build` stage and we can get the executable artifact from that (I think an [artifact](https://docs.gitlab.com/ee/ci/yaml/#artifacts) would be the best way to go about passing the executable from one stage to another). 
   We'd like to then have this dedicated job that just uploads the binary to the S3 bucket. This also means we can isolate where issues in our CI might be occurring. Or (hypothetically) we could later add an additional upload job that uploads to azure (or other places down the line) and doing that wouldn't require us to change the binary build job and we also wouldn't have to alter our aws copy job either.

Reply:
@JoseD92 did you have any questions on the changes above?